report/metrics.go

```diff
@@ -109,83 +134,66 @@ func last(t1, t2 time.Time) time.Time {
 	return t2
 }
 
-// concat returns a new list formed by adding each element of acc to curr
-// acc or curr can be nil.
-func concat(acc []interface{}, curr ps.List) ps.List {
-	if curr == nil {
-		curr = ps.NewList()
-	}
-	for i := len(acc) - 1; i >= 0; i-- {
-		curr = curr.Cons(acc[i])
-	}
-	return curr
-}
-
-// Add returns a new Metric with (t, v) added to its Samples. Add is the only
-// valid way to grow a Metric.
-func (m Metric) Add(t time.Time, v float64) Metric {
-	// Find the first element which is before you element, and insert
-	// your new element in the list.  NB we want to dedupe entries with
-	// equal timestamps.
-	// This should be O(1) to insert a latest element, and O(n) in general.
-	curr, acc := m.Samples, make([]interface{}, 0, m.Len()+1)
-	for {
-		if curr == nil || curr.IsNil() {
-			break
-		}
+// Merge combines the two Metrics and returns a new result.
+func (m Metric) Merge(other Metric) Metric {
 
-		currSample := curr.Head().(Sample)
-		if currSample.Timestamp.Equal(t) {
-			curr = curr.Tail()
-			break
+	// Optimize the empty and non-overlapping case since they are very common
+	switch {
+	case len(m.Samples) == 0:
+		return other.Copy()
+	case len(other.Samples) == 0:
+		return m.Copy()
```

Comment:
> I did it lest of an external modification [...] But in that case an explicit copy is made.

Any such modification *w/o* copying would break the contract we agreed in #1728 is expected of `Merge()`.